support flow description YAML

gituser173 opened the pull request:
Fixes #505
 
Signed-off-by: Claudiu Ban claudiu-laurian.ban@hpe.com

gituser173 commented:
@Bonczidai please review(I will add the unit tests after your review)

Bonczidai commented on cloudslang-all/src/main/java/io/cloudslang/lang/api/Slang.java:
```diff
@@ -31,7 +32,14 @@
+     * @param source the CloudSlang source containing the flow
+     * @return the metadata of the flow
+     */
+    Metadata extractMetadata(SlangSource source);
```
I'd put this method first for logical order

Bonczidai commented on cloudslang-cli/src/main/java/io/cloudslang/lang/cli/utils/MetadataHelperImpl.java:
```diff
@@ -0,0 +1,30 @@
+    private Slang slang;
+
+    @Override
+    public Metadata extractMetadata(File file) throws IOException {
```
Y declare IOException here?

Bonczidai commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java:
```diff
@@ -0,0 +1,79 @@
+ * Date: 1/11/2016
+ */
+@Component
+public class YAMLMetadataExtractorImpl implements MetadataExtractor {
```
according to convention, class name should be MetadataExtractorImpl

Bonczidai commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java:
```diff
@@ -0,0 +1,79 @@
+            String line = reader.readLine();
+            while (line != null) {
+                if (line.startsWith("##")) {
+                    line = line.substring(3);
```
does it work for such line:
```
##
```
?

Bonczidai commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java:
```diff
@@ -0,0 +1,79 @@
+                line = reader.readLine();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Read issue(should not happen)", e);
```
error message should be more descriptive ~ Error processing metadata, error extracting metadata from <source name>

Bonczidai commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java:
```diff
@@ -0,0 +1,79 @@
+        }
+        return sb.toString();
+    }
+}
```
you extract `##`, `3` and exception message as consts

Bonczidai commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/Metadata.java:
```diff
@@ -0,0 +1,47 @@
+        return inputs;
+    }
+
+    public void setInputs(Map<String, String> inputs) {
```
is there any reason for setter here? better add a constructor with params

gituser173 replied:
I find it easier to read separate setters than a constructor/method with many parameters

Bonczidai commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/MetadataYAML.java:
```diff
@@ -0,0 +1,48 @@
+        return outputs;
+    }
+
+    public void setOutputs(List<Map<String, String>> outputs) {
```
same for setters here

gituser173 replied:
the snakeyaml library doesn't seem to work when I remove the setters

Bonczidai commented:
little redesign suggestion in compiler package:

it has three phases: parse (you get the doc as maps, lists, strings etc.) -> model (you get objects of your choice) -> compile to Execution Plan (not affected here)

- move `MetadatYAML` to `parser.model` - rename ~ `ParsedMetadata`
- `Metadata` stays in `modeller.model`
- break `MetadataExtractor` in three:
  - `MetadataParser` returns `MetadatYAML` in `parser` package
  - `MetadataModeller` returns `Metadata` in `modeller.model` package
  - `MetadataExtractor` wires the two and exposes the interface 

WDYT?

Bonczidai commented:
:ok_hand:  good work

please handle comments and add tests

gituser173 commented:
about the redesign suggestions:
I would not break the MetadataExtractor in three unless we have a use for MetadataParser that returns MetadatYAML in parser package in other places. MetadataModeler is just a converter from MetadataYAML to Metadata pojos, don't think it would be of any use anywhere else, I would keep those two small private methods inside MetadataExtractor. 

Moved and renamed MetadataYAML to ParsedMetada :)

gituser173 commented:
@SamMarkowitz 
![capture](https://cloud.githubusercontent.com/assets/12094982/12323566/0d70861c-bac5-11e5-9282-64dd7993378e.PNG)

SamMarkowitz commented:
Hmm. I think all the bracketing, commas, and quotes of what I assume is a JSON object is a bit confusing. Can all that be stripped away?

Also, there are "\n"s where in the documentation there are new lines.

gituser173 commented:
@SamMarkowitz 
![capture](https://cloud.githubusercontent.com/assets/12094982/12353678/7c8df280-bb99-11e5-84aa-6bce4e3366ae.PNG)

Bonczidai commented:
two Qs:
- why using `.` separator in file names instead of `_`, we should follow our conventions IMHO
- how does the CLI output look like when inspecting an invalid description - please put a screenshot here

SamMarkowitz commented:
Looks great! For the inputs can we get the dash after the pipe to show up on the next line or get rid of the dashes altogether (which I think is the last way we had it in issue 505).

```
inputs:
  json_input:
    JSON data input
    example: some example   
```

gituser173 commented:
@Bonczidai 
![capture](https://cloud.githubusercontent.com/assets/12094982/12386029/6f6b540c-bdca-11e5-9da2-4441924e0a70.PNG)
